src/operator/numpy/np_where_op-inl.h

```diff
@@ -0,0 +1,218 @@
+
+using namespace mshadow;
+
+template<int ndim>
+struct numpy_where_kernel {
+  template<typename CType, typename DType>
+  MSHADOW_XINLINE static void Map(index_t base, OpReqType req, const Shape<ndim> &cstride,
+                                  const Shape<ndim> &xstride, const Shape<ndim> &ystride,
+                                  const Shape<ndim> &oshape, CType *datac, DType *datax,
+                                  DType *datay, DType *out) {
+    Shape<ndim> coord = mxnet_op::unravel(base, oshape);
+    auto cidx = static_cast<index_t>(mxnet_op::dot(coord, cstride));
+    auto xidx = static_cast<index_t>(mxnet_op::dot(coord, xstride));
+    auto yidx = static_cast<index_t>(mxnet_op::dot(coord, ystride));
+    KERNEL_ASSIGN(out[base], req, datac[cidx] != CType(0) ? datax[xidx] : datay[yidx]);
+  }
+};
+
+template<int ndim, bool is_left>
+struct numpy_where_backward_kernel {
+  template<typename CType, typename DType>
+  MSHADOW_XINLINE static void Map(index_t base, OpReqType req,
+                                  const Shape<ndim> &cstride, const Shape<ndim> &oshape,
+                                  CType *datac, DType *datao, DType *grad) {
+    Shape<ndim> coord = mxnet_op::unravel(base, oshape);
+    auto cidx = static_cast<index_t>(mxnet_op::dot(coord, cstride));
+    if (is_left) {
+      KERNEL_ASSIGN(grad[base], req, datac[cidx] != CType(0) ? datao[base] : DType(0));
+    } else {
+      KERNEL_ASSIGN(grad[base], req, datac[cidx] == CType(0) ? datao[base] : DType(0));
+    }
+  }
+};
+
+template<typename xpu>
+inline void NumpyWhereOpForward(const nnvm::NodeAttrs& attrs,
+                                const OpContext& ctx,
+                                const std::vector<TBlob>& inputs,
+                                const std::vector<OpReqType>& req,
+                                const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 1U);
+  if (outputs[0].shape_.Size() == 0U) return;  // zero-size tensor
+  CHECK_LE(outputs[0].shape_.ndim(), broadcast::MAX_DIM);
+
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  std::vector<Shape<broadcast::MAX_DIM>> in_strides;
+  in_strides.resize(3);
+  for (int i = 0; i < 3; ++i) {
+    TShape expanded_ishape(broadcast::MAX_DIM, 1);
+    const TShape& ishape = inputs[i].shape_;
+    const int ndim_delta = expanded_ishape.ndim() - ishape.ndim();
+    for (int j = 0; j < ishape.ndim(); ++j) {
+      expanded_ishape[j + ndim_delta] = ishape[j];
+    }
+    in_strides[i] = mxnet_op::calc_stride(expanded_ishape.get<broadcast::MAX_DIM>());
+  }
+  TShape expanded_oshape(broadcast::MAX_DIM, 1);
+  const int ndim_delta = expanded_oshape.ndim() - outputs[0].shape_.ndim();
+  for (int j = 0; j < outputs[0].shape_.ndim(); ++j) {
+    expanded_oshape[j + ndim_delta] = (outputs[0].shape_)[j];
+  }
+  Shape<broadcast::MAX_DIM> oshape = expanded_oshape.get<broadcast::MAX_DIM>();
+  MSHADOW_TYPE_SWITCH_WITH_BOOL(outputs[0].type_flag_, DType, {
```

Comment:
Consider do:
```
const TBlob& cond = inputs[0];
const TBlob& x = inputs[1];
const TBlob& y = inputs[2];
const TBlob& out = outputs[0];
```
and substitute all the `inputs[i]` `outputs[j]` to improve readability.